palette/src/luma/relative_contrast.rs

```diff
@@ -0,0 +1,57 @@
+use from_f64;
+use FloatComponent;
+
+/// A trait for calculating relative contrast between two colors.
+///
+/// W3C's Web Content Accessibility Guidelines (WCAG) 2.1 suggest a method
+/// to calculate accessible contrast ratios of text and background colors
+/// for those with low vision or color deficiencies, and for contrast of user
+/// interface graphics. The possible range of ratios is from 1:1 to 21:1.
+/// There is a Success Criterion for Contrast (Minimum) and a Success Criterion
+/// for Contrast (Enhanced), SC 1.4.3 and SC 1.4.6 respectively, which are
+/// concerned with text and images of text. SC 1.4.11 is a Success Criterion for
+/// "non-text contrast" such as user interface components and other graphics.
+/// The relative contrast is calculated by `(L1 + 0.05) / (L2 + 0.05)`, where
+/// `L1` is the luminance of the brighter color and `L2` is the luminance of the
+/// darker color both in sRGB linear space. Higher contrast ratio is generally
+/// desireable.
+///
+/// For more details of the guidelines:
+///
+/// [Success Criterion 1.4.3 Contrast (Minimum) (Level AA)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum)
+///
+/// [Success Criterion 1.4.6 Contrast (Enhanced) (Level AAA)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-enhanced)
+///
+/// [Success Criterion 1.4.11 Non-text Contrast (Level AA)](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html)
+pub trait RelativeContrast {
+    /// The return type for the contrast ratio.
+    type Scalar: FloatComponent;
+
+    /// Calculate contrast ratio between two colors.
+    fn get_contrast_ratio(&self, other: &Self) -> Self::Scalar;
+    /// Verify the contrast between two colors satisfies SC 1.4.3. Contrast
+    /// is at least 4.5:1 (Level AA).
+    fn is_min_contrast_text(&self, other: &Self) -> bool {
+        self.get_contrast_ratio(other) >= from_f64(4.5)
+    }
+    /// Verify the contrast between two colors satisfies SC 1.4.3 for large
+    /// text. Contrast is at least 3:1 (Level AA).
+    fn is_min_contrast_large_text(&self, other: &Self) -> bool {
+        self.get_contrast_ratio(other) >= from_f64(3.0)
+    }
+    /// Verify the contrast between two colors satisfies SC 1.4.6. Contrast
+    /// is at least 7:1 (Level AAA).
+    fn is_enhanced_contrast_text(&self, other: &Self) -> bool {
+        self.get_contrast_ratio(other) >= from_f64(7.0)
+    }
+    /// Verify the contrast between two colors satisfies SC 1.4.6 for large
+    /// text. Contrast is at least 4.5:1 (Level AAA).
+    fn is_enhanced_contrast_large_text(&self, other: &Self) -> bool {
+        self.is_min_contrast_text(other)
+    }
+    /// Verify the contrast between two colors satisfies SC 1.4.11 for graphical
+    /// objects. Contrast is at least 3:1 (Level AA).
+    fn is_non_text_contrast(&self, other: &Self) -> bool {
```

Comment:
Or `is_min_contrast_non_text`? Or just `is_min_contrast`? Maybe `is_min_contrast_graphics`?

Besides, isn't it more correct to say that the color _has_ some contrast? As in `has_min_contrast(_for)_text`. I'm not a native speaker, so I may definitely be wrong.

Reply:
Any of those names are fine, I actually prefer `is_min_contrast_graphics`. The heading for that recommendation is called `Success Criterion 1.4.11 Non-text Contrast`. My reasoning for not including `min_contrast` in that particular function name was because `Success Criterion 1.4.3 Contrast (Minimum)` has minimum in the name and 1.4.11 is a separate criterion which does not. There is no minimum or enhanced contrast for graphical/non-text objects, only the recommendation of 3:1, so I didn't want it to be confused with 1.4.3. I'm not attached to any of the names so changing that is fine as 1.4.11 does fall in the same Level AA of Compliance as 1.4.3 which is the "minimum" for contrast levels (but still not Level A which seems more like a fundamental level for content).

I chose `is_*_contrast*` thinking of it like "is this (compliant with) recommendation criteria". I think `has_min_contrast` is more clear about the idea that "this color has a minimum contrast with other color" so `has_min_contrast_graphics` might be the way to go. I think I should rename the other functions now that I think about it.

Reply:
Sounds good. I thought it sounded weird to think of it as "is this color a (non-text) contrast"? Other than that, I'm trying to align them in some way. It's probably ok to go a step further and use `_min_`, because that seem to be the intention. As in "anything under this is not recommend".